### PR TITLE
Backports release 1.9

### PR DESCRIPTION
**Pull request by KristofferC**
- [x] [Revert "Resolve artifact ambiguities using shortest match"](https://github.com/JuliaLang/julia/pull/49500/commits/6f2f44e602a3b9641a1538605b5e0f7432e0f4cd) 

Fixes (on release-1.9) https://github.com/JuliaLang/julia/issues/49491

cc @giordano, @staticfloat

### CODE REVIEW

File: base/binaryplatforms.jl [hunk 1]
```diff
@@ -1066,20 +1066,12 @@ function select_platform(download_info::Dict, platform::AbstractPlatform = HostP
...
-    # the candidate matches by their triplets and take the last one, so as to
-    # generally choose the latest release (e.g. a `libgfortran5` tarball over a
-    # `libgfortran3` tarball).
+    # At this point, we may have multiple possibilities.  E.g. if, in the future,
+    # Julia can be built without a direct dependency on libgfortran, we may match
+    # multiple tarballs that vary only within their libgfortran ABI.  To narrow it
+    # down, we just sort by triplet, then pick the last one.  This has the effect
+    # of generally choosing the latest release (e.g. a `libgfortran5` tarball
+    # rather than a `libgfortran3` tarball)
     p = last(sort(ps, by = p -> triplet(p)))
```
Comment (StefanKarpinski) [inline base/binaryplatforms.jl:1075]: Gonna be a bad surprise when we encounter `libgfortran10`.

**Comment by StefanKarpinski** [comment]
@staticfloat, this change really needs to do a natural sort, not just here but on master as well.

**Comment by giordano** [comment]
The issue is still relevant on `master`, this is only avoiding troubles on 1.9.

**Comment by StefanKarpinski** [comment]
Do we have an issue for it or should I open one?

**Comment by giordano** [comment]
#49491, the issue mentioned above

File: base/binaryplatforms.jl (hunk 1 above)
Comment (staticfloat) [reply base/binaryplatforms.jl:1075]: This secondary sorting only exists to get rid of dict-order selection and use something more deterministic when there is an incomplete specification.  This should never happen with JLLs except in very rare circumstances (the only such circumstance I know of right now is the memory sanitizer builds, which is what these PRs are trying to address).

I've opened a PR to implement a better matching algorithm here: https://github.com/JuliaLang/julia/pull/49502

**Pull request merged**